contracts/oracle/CollateralizationMemoizer.sol

```diff
@@ -0,0 +1,204 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.4;
+
+import "./IOracle.sol";
+import "./ICollateralizationOracle.sol";
+import "../refs/CoreRef.sol";
+
+interface IPausable {
+    function paused() external view returns (bool);
+}
+
+/// @title Fei Protocol's Collateralization Oracle
+/// @author eswak
+/// @notice Reads a list of PCVDeposit that report their amount of collateral
+///         and the amount of protocol-owned FEI they manage, to deduce the
+///         protocol-wide collateralization ratio.
+contract CollateralizationMemoizer is IOracle, CoreRef {
+    using Decimal for Decimal.D256;
+
+    // ----------- Events -----------
+
+    event CachedValueUpdate(
+        address from,
+        uint256 indexed protocolControlledValue,
+        uint256 indexed userCirculatingFei,
+        int256 indexed protocolEquity
+    );
+
+    // ----------- Properties -----------
+
+    /// @notice address of the CollateralizationOracle to memoize
+    address public collateralizationOracle;
+
+    /// @notice cached value of the Protocol Controlled Value
+    uint256 public cachedProtocolControlledValue;
+    /// @notice cached value of the User Circulating FEI
+    uint256 public cachedUserCirculatingFei;
+    /// @notice cached value of the Protocol Equity
+    int256 public cachedProtocolEquity;
+
+    /// @notice timestamp of the last read operation on the CollateralizationOracle
+    uint256 public lastUpdate;
+    /// @notice validity duration of the cached value
+    uint256 public validityDuration;
+    /// @notice deviation threshold to consider cached values outdated, in basis
+    ///         points (base 10_000)
+    uint256 public deviationThreshold;
+
+    // ----------- Constructor -----------
+
+    /// @notice CollateralizationOracle constructor
+    /// @param _core Fei Core for reference
+    constructor(
+        address _core
+    ) CoreRef(_core) {}
+
+    // ----------- IOracle override methods -----------
+    /// @notice update reading of the CollateralizationOracle
+    function update() external override whenNotPaused {
+        // fetch a fresh round of information
+        (
+            uint256 _protocolControlledValue,
+            uint256 _userCirculatingFei,
+            int256 _protocolEquity,
+            bool _validityStatus
+        ) = ICollateralizationOracle(collateralizationOracle).pcvStats();
+
+        // only update if valid
+        require(_validityStatus, "CollateralizationMemoizer: CollateralizationOracle is invalid");
+
+        // Update cached variables
+        _setCache(_protocolControlledValue, _userCirculatingFei, _protocolEquity);
+    }
+
+    // @notice returns true if the cached values are outdated.
+    function isOutdated() external override view returns (bool outdated) {
```

Comment:
I think this function should use early returning liberally to simplify the logic

Reply:
This method should split in two, isOutdated returning the time condition only and a second called isExceededDeviationThreshold

Reply:
I did not use early returns, since the function is intended to be called offchain only, I kept the code as simple as possible without too much branching